File: src/features/diff/entities/diff.ts
import { CommitChangedFileSummaryDto, DiffFileDto } from "../types/diff-dto";

export type DiffChangeType = "modified" | "added" | "deleted" | "renamed" | "binary";
export type DiffLineType = "context" | "added" | "removed";

export type DiffLine = {
  type: DiffLineType;
  content: string;
};

export type DiffHunk = {
  header: string;
  lines: DiffLine[];
};

export type DiffFile = {
  path: string;
  oldPath: string | null;
  changeType: DiffChangeType;
  isBinary: boolean;
  hunks: DiffHunk[];
};

export function toDiffFile(diffFileDto: DiffFileDto): DiffFile {
  return {
    path: diffFileDto.path,
    oldPath: diffFileDto.oldPath ?? null,
    changeType: mapChangeType(diffFileDto.changeType),
    isBinary: diffFileDto.isBinary,
    hunks: diffFileDto.hunks.map((hunk) => ({
      header: hunk.header,
      lines: hunk.lines.map((line) => ({
        type: mapLineType(line.type),
        content: line.content,
      })),
    })),
  };
}

/** Summary row from `list_commit_changed_files` (no hunks until per-file diff loads). */
export function commitChangedSummaryToDiffFile(summary: CommitChangedFileSummaryDto): DiffFile {
  return {
    path: summary.path,
    oldPath: summary.oldPath ?? null,
    changeType: mapChangeType(summary.changeType),
    isBinary: summary.isBinary,
    hunks: [],
  };
}

function mapChangeType(changeType: DiffFileDto["changeType"]): DiffChangeType {
  switch (changeType) {
    case "added":
      return "added";
    case "deleted":
      return "deleted";
    case "renamed":
      return "renamed";
    case "binary":
      return "binary";
    case "modified":
    default:
      return "modified";
  }
}

function mapLineType(lineType: "context" | "added" | "removed"): DiffLineType {
  switch (lineType) {
    case "added":
      return "added";
    case "removed":
      return "removed";
    case "context":
    default:
      return "context";
  }
}

File: src-tauri/src/interfaces/dto/diff.rs
use serde::{Deserialize, Serialize};

use super::app_error::AppErrorDto;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DiffChangeTypeDto {
    Modified,
    Added,
    Deleted,
    Renamed,
    Binary,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DiffLineTypeDto {
    Context,
    Added,
    Removed,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiffLineDto {
    pub r#type: DiffLineTypeDto,
    pub content: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiffHunkDto {
    pub header: String,
    pub lines: Vec<DiffLineDto>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiffFileDto {
    pub path: String,
    pub old_path: Option<String>,
    pub change_type: DiffChangeTypeDto,
    pub is_binary: bool,
    pub hunks: Vec<DiffHunkDto>,
}

#[derive(Debug, Clone, Copy, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub enum WorkingDiffScopeDto {
    #[default]
    Unstaged,
    Staged,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetWorkingDiffInputDto {
    pub repository_path: String,
    pub file_path: String,
    /// When omitted, defaults to unstaged (`git diff`).
    #[serde(default)]
    pub scope: WorkingDiffScopeDto,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetWorkingDiffResponseDto {
    pub files: Option<Vec<DiffFileDto>>,
    pub error: Option<AppErrorDto>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetCommitDiffInputDto {
    pub repository_path: String,
    pub commit_hash: String,
    pub file_path: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetCommitDiffResponseDto {
    pub files: Option<Vec<DiffFileDto>>,
    pub error: Option<AppErrorDto>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListCommitChangedFilesInputDto {
    pub repository_path: String,
    pub commit_hash: String,
}

/// Paths touched by a commit without unified-diff hunks (for lazy per-file loading).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitChangedFileSummaryDto {
    pub path: String,
    pub old_path: Option<String>,
    pub change_type: DiffChangeTypeDto,
    pub is_binary: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListCommitChangedFilesResponseDto {
    pub files: Option<Vec<CommitChangedFileSummaryDto>>,
    pub error: Option<AppErrorDto>,
}
